platform/mbed_error.h

```diff
@@ -23,13 +22,782 @@
+ * @return                  0 or ERROR_SUCCESS.
+ *                          ERROR_INVALID_ARGUMENT if called with invalid error status/codes
+ *
+ * @code
+ * 
+ * SET_ERROR_FATAL( ERROR_MUTEX_LOCK_FAILED, "MyDriver: Can't lock driver Mutex", &my_mutex )
+ *
+ * @endcode
+ * @note The macro calls set_error_fatal API with filename and line number info without caller explicitly passing them.
+*        Since this macro is a wrapper for set_error_fatal API callers should process the return value from this macro which is the return value from calling set_error_fatal API. 
+ *
+ */
+#ifdef MBED_CONF_ERROR_FILENAME_CAPTURE_ENABLED
+    #define SET_ERROR_FATAL( error_status, error_msg, error_value )       set_error_fatal( error_status, (const char *)error_msg, (uint32_t)error_value, (const char *)MBED_FILENAME, __LINE__ )
+#else
+    #define SET_ERROR_FATAL( error_status, error_msg, error_value )       set_error_fatal( error_status, (const char *)error_msg, (uint32_t)error_value, NULL, 0 )
+#endif         
+
+//Error Type definition
+/** MbedErrorType definition
+ *  @note
+ *  This enumeration defines the Error types supported. The value of these enum values will be encoded into MbedErrorStatus TYPE field.\n
+ *      See MbedErrorStatus description for more info.\n
+ *         ERROR_TYPE_SYSTEM - Used to indicate that the error status is of System defined Error type.\n
+ *         ERROR_TYPE_CUSTOM - Used to indicate that the error status is of Custom defined Error type.\n 
+ *         ERROR_TYPE_POSIX  - Used to indicate that the error status is of Posix error type.\n 
+ *
+ */
+typedef enum _MbedErrorType
+{
+    ERROR_TYPE_SYSTEM = 0,
+    ERROR_TYPE_CUSTOM = 1,
+    //2 is reserved
+    //Use 3 for POSIX because we are mapping -1 to -255 to POSIX error codes
+    //and thus we must use 3 to match the type bits in error status representation which are from 0xFFFFFFFF to 0xFFFFFF00
+    ERROR_TYPE_POSIX = 3
+} MbedErrorType;        
+
+//Entity type/id definitions
+/** MbedEntityType definition
+ * @note
+ * This enumeration defines the Entity types. The value of these enum values will be encoded into MbedErrorStatus ENTITY field.\n\n
+ *      See MbedErrorStatus description for more info.\n
+ *          ENTITY_UNKNOWN - This entity type can be used if caller of the set_error/set_error_fatal doesn't know who is the actual originator of the error.\n
+ *                       Other entity values can be used to provide more info on who/where the error originated from.\n\n
+ *                       For example, if I2C driver is the component originating the error you can use ENTITY_DRIVER_I2C to provide more info.\n
+ *                       Its used in call to MAKE_ERROR/MAKE_SYSTEM_ERROR/MAKE_CUSTOM_ERROR macros.\n
+ *
+ * @code
+ *      Example: MbedErrorStatus i2c_driver_error = MAKE_ERROR( ENTITY_DRIVER_I2C, ERROR_CONFIG_UNSUPPORTED );
+ * @endcode
+ * 
+ * @note
+ *  \n Below are the entity code mappings:\n
```

Comment:
In fact I tried it, its not getting rendered where I wanted to be. But Ill try again if it can be moved down along with actual definition. I also thought we may provide more detailed info on these error code later and so making them separate makes sense.